Flaky AS NATS test

<!--
Thanks for submitting a bug report. Please fill the template below,
otherwise we will not be able to process this bug report.
-->

#### Summary
<!-- Summarize the problem in a few sentences: -->

I encountered a flaky test in the AS NATS package.

#### What do you see now?
<!-- Please paste terminal output, upload logs (as .txt) or upload screenshots. -->

```
--- FAIL: TestOpenConnection (2.08s)
    doc.go:66: 
        /home/travis/gopath/src/go.thethings.network/lorawan-stack/pkg/applicationserver/io/pubsub/provider/nats/provider_test.go:164
        Expected: nil
        Actual:   'context deadline exceeded'
    doc.go:66: 
        /home/travis/gopath/src/go.thethings.network/lorawan-stack/pkg/applicationserver/io/pubsub/provider/nats/provider_test.go:165
        Expected '<nil>' to NOT be nil (but it was)!
    --- FAIL: TestOpenConnection/Downstream (2.05s)
        --- FAIL: TestOpenConnection/Downstream/ValidPush (2.05s)
panic: runtime error: invalid memory address or nil pointer dereference [recovered]
	panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x10 pc=0x136fc6d]
goroutine 40 [running]:
testing.tRunner.func1.1(0x14882e0, 0x25672b0)
	/home/travis/.gimme/versions/go1.14.4.linux.amd64/src/testing/testing.go:940 +0x2f5
testing.tRunner.func1(0xc000558360)
	/home/travis/.gimme/versions/go1.14.4.linux.amd64/src/testing/testing.go:943 +0x3f9
panic(0x14882e0, 0x25672b0)
	/home/travis/.gimme/versions/go1.14.4.linux.amd64/src/runtime/panic.go:969 +0x166
go.thethings.network/lorawan-stack/v3/pkg/applicationserver/io/pubsub/provider/nats_test.TestOpenConnection.func1.1(0xc000558360)
	/home/travis/gopath/src/go.thethings.network/lorawan-stack/pkg/applicationserver/io/pubsub/provider/nats/provider_test.go:167 +0x2ed
testing.tRunner(0xc000558360, 0xc00053e330)
	/home/travis/.gimme/versions/go1.14.4.linux.amd64/src/testing/testing.go:991 +0xdc
created by testing.(*T).Run
	/home/travis/.gimme/versions/go1.14.4.linux.amd64/src/testing/testing.go:1042 +0x357
FAIL	go.thethings.network/lorawan-stack/v3/pkg/applicationserver/io/pubsub/provider/nats	2.108s
```

--- a/pkg/applicationserver/io/pubsub/provider/mqtt/provider_test.go
+++ b/pkg/applicationserver/io/pubsub/provider/mqtt/provider_test.go
@@ -178,20 +178,20 @@ func TestOpenConnection(t *testing.T) {
 			},
 		},
 	} {
-		client := utc.createClient(t, a)
-		defer client.Disconnect(uint(timeout / time.Millisecond))
-
-		unsubscribe := func(t *testing.T, topic string) {
-			token := client.Unsubscribe(topic)
-			if !token.WaitTimeout(timeout) {
-				t.Fatal("Unsubscribe timeout")
-			}
-			if !a.So(token.Error(), should.BeNil) {
-				t.FailNow()
-			}
-		}
-
 		t.Run(utc.name, func(t *testing.T) {
+			client := utc.createClient(t, a)
+			defer client.Disconnect(uint(timeout / time.Millisecond))
+
+			unsubscribe := func(t *testing.T, topic string) {
+				token := client.Unsubscribe(topic)
+				if !token.WaitTimeout(timeout) {
+					t.Fatal("Unsubscribe timeout")
+				}
+				if !a.So(token.Error(), should.BeNil) {
+					t.FailNow()
+				}
+			}
+
 			pb.Provider = utc.provider
 
 			conn, err := impl.OpenConnection(ctx, pb)

--- a/pkg/applicationserver/io/pubsub/provider/mqtt/provider_test.go
+++ b/pkg/applicationserver/io/pubsub/provider/mqtt/provider_test.go
@@ -178,6 +178,19 @@ func TestOpenConnection(t *testing.T) {
 			},
 		},
 	} {
+		client := utc.createClient(t, a)
+		defer client.Disconnect(uint(timeout / time.Millisecond))
+
+		unsubscribe := func(t *testing.T, topic string) {
+			token := client.Unsubscribe(topic)
+			if !token.WaitTimeout(timeout) {
+				t.Fatal("Unsubscribe timeout")
+			}
+			if !a.So(token.Error(), should.BeNil) {
+				t.FailNow()
+			}
+		}
+
 		t.Run(utc.name, func(t *testing.T) {
 			pb.Provider = utc.provider
 
@@ -222,11 +235,6 @@ func TestOpenConnection(t *testing.T) {
 				} {
 					t.Run(tc.name, func(t *testing.T) {
 						a := assertions.New(t)
-						ctx, cancel := context.WithTimeout(ctx, timeout)
-						defer cancel()
-
-						client := utc.createClient(t, a)
-						defer client.Disconnect(uint(timeout / time.Millisecond))
 
 						token := client.Publish(tc.topicName, 2, false, "foobar")
 						if !token.WaitTimeout(timeout) {
@@ -236,12 +244,16 @@ func TestOpenConnection(t *testing.T) {
 							t.FailNow()
 						}
 
+						ctx, cancel := context.WithTimeout(ctx, timeout)
+						defer cancel()
+
 						msg, err := tc.subscription.Receive(ctx)
 						if tc.expectMessage {
 							a.So(err, should.BeNil)
-							a.So(msg, should.NotBeNil)
 
-							a.So(msg.Body, should.Resemble, []byte("foobar"))
+							if a.So(msg, should.NotBeNil) {
+								a.So(msg.Body, should.Resemble, []byte("foobar"))
+							}
 						} else if err == nil {
 							t.Fatal("Unexpected message received")
 						}
@@ -301,9 +313,6 @@ func TestOpenConnection(t *testing.T) {
 					t.Run(tc.name, func(t *testing.T) {
 						a := assertions.New(t)
 
-						client := utc.createClient(t, a)
-						defer client.Disconnect(uint(timeout / time.Millisecond))
-
 						upCh := make(chan paho_mqtt.Message, 10)
 						defer close(upCh)
 						token := client.Subscribe(tc.topicName, 2, func(_ paho_mqtt.Client, msg paho_mqtt.Message) {
@@ -315,6 +324,10 @@ func TestOpenConnection(t *testing.T) {
 						if !a.So(token.Error(), should.BeNil) {
 							t.FailNow()
 						}
+						defer unsubscribe(t, tc.topicName)
+
+						ctx, cancel := context.WithTimeout(ctx, timeout)
+						defer cancel()
 
 						err = tc.topic.Send(ctx, &pubsub.Message{
 							Body: []byte("foobar"),
@@ -325,8 +338,9 @@ func TestOpenConnection(t *testing.T) {
 						case <-time.After(timeout):
 							t.Fatal("Expected message never arrived")
 						case msg := <-upCh:
-							a.So(msg, should.NotBeNil)
-							a.So(msg.Payload(), should.Resemble, []byte("foobar"))
+							if a.So(msg, should.NotBeNil) {
+								a.So(msg.Payload(), should.Resemble, []byte("foobar"))
+							}
 						}
 					})
 				}

--- a/pkg/applicationserver/io/pubsub/provider/nats/provider_test.go
+++ b/pkg/applicationserver/io/pubsub/provider/nats/provider_test.go
@@ -46,6 +46,10 @@ func TestOpenConnection(t *testing.T) {
 	a.So(natsServer, should.NotBeNil)
 	defer natsServer.Shutdown()
 
+	natsClient, err := nats_client.Connect("nats://localhost:4123")
+	a.So(err, should.BeNil)
+	defer natsClient.Close()
+
 	pb := &ttnpb.ApplicationPubSub{
 		ApplicationPubSubIdentifiers: ttnpb.ApplicationPubSubIdentifiers{
 			ApplicationIdentifiers: ttnpb.ApplicationIdentifiers{
@@ -116,8 +120,10 @@ func TestOpenConnection(t *testing.T) {
 		conn, err := impl.OpenConnection(ctx, pb)
 		a.So(conn, should.NotBeNil)
 		a.So(err, should.BeNil)
-
 		defer conn.Shutdown(ctx)
+
+		// Wait for subscriptions to connect, since they are not synchronous.
+		time.Sleep(timeout)
 
 		t.Run("Downstream", func(t *testing.T) {
 			for _, tc := range []struct {
@@ -152,22 +158,18 @@ func TestOpenConnection(t *testing.T) {
 				},
 			} {
 				t.Run(tc.name, func(t *testing.T) {
-					ctx, cancel := context.WithTimeout(ctx, timeout)
-					defer cancel()
-
-					natsClient, err := nats_client.Connect("nats://localhost:4123")
-					a.So(err, should.BeNil)
-					defer natsClient.Close()
-
 					err = natsClient.Publish(tc.subject, []byte("foobar"))
 					a.So(err, should.BeNil)
+
+					ctx, cancel := context.WithTimeout(ctx, timeout)
+					defer cancel()
 
 					msg, err := tc.subscription.Receive(ctx)
 					if tc.expectMessage {
 						a.So(err, should.BeNil)
-						a.So(msg, should.NotBeNil)
-
-						a.So(msg.Body, should.Resemble, []byte("foobar"))
+						if a.So(msg, should.NotBeNil) {
+							a.So(msg.Body, should.Resemble, []byte("foobar"))
+						}
 					} else if err == nil {
 						t.Fatal("Unexpected message received")
 					}
@@ -225,10 +227,6 @@ func TestOpenConnection(t *testing.T) {
 				},
 			} {
 				t.Run(tc.name, func(t *testing.T) {
-					natsClient, err := nats_client.Connect("nats://localhost:4123")
-					a.So(err, should.BeNil)
-					defer natsClient.Close()
-
 					upCh := make(chan *nats_client.Msg, 10)
 					defer close(upCh)
 
@@ -242,6 +240,9 @@ func TestOpenConnection(t *testing.T) {
 					// was actually opened.
 					time.Sleep(timeout)
 
+					ctx, cancel := context.WithTimeout(ctx, timeout)
+					defer cancel()
+
 					err = tc.topic.Send(ctx, &pubsub.Message{
 						Body: []byte("foobar"),
 					})
@@ -251,8 +252,9 @@ func TestOpenConnection(t *testing.T) {
 					case <-time.After(timeout):
 						t.Fatal("Expected message never arrived")
 					case msg := <-upCh:
-						a.So(msg, should.NotBeNil)
-						a.So(msg.Data, should.Resemble, []byte("foobar"))
+						if a.So(msg, should.NotBeNil) {
+							a.So(msg.Data, should.Resemble, []byte("foobar"))
+						}
 					}
 				})
 			}

